patches/server/0846-Add-PlayerInsertLecternBookEvent.patch

```diff
@@ -0,0 +1,50 @@
+Date: Sat, 29 Aug 2020 21:55:50 +0200
+Subject: [PATCH] Add PlayerInsertLecternBookEvent
+
+
+diff --git a/src/main/java/net/minecraft/world/level/block/LecternBlock.java b/src/main/java/net/minecraft/world/level/block/LecternBlock.java
+index 4d19c5f4653831a7c02e691e0585bf45d1d0ee64..fcf0ae3d2767f3a61d80252e794b11f77eea56b8 100644
+--- a/src/main/java/net/minecraft/world/level/block/LecternBlock.java
++++ b/src/main/java/net/minecraft/world/level/block/LecternBlock.java
+@@ -135,6 +135,7 @@ public class LecternBlock extends BaseEntityBlock {
+         if (!(Boolean) state.getValue(LecternBlock.HAS_BOOK)) {
+             if (!world.isClientSide) {
+                 LecternBlock.placeBook(player, world, pos, state, stack);
++                return ((LecternBlockEntity) world.getBlockEntity(pos)).hasBook();
+             }
+ 
+             return true;
+@@ -149,7 +150,8 @@ public class LecternBlock extends BaseEntityBlock {
+         if (tileentity instanceof LecternBlockEntity) {
+             LecternBlockEntity tileentitylectern = (LecternBlockEntity) tileentity;
+ 
+-            tileentitylectern.setBook(stack.split(1));
++            tileentitylectern.setBook(stack.split(1), player); // Paper
++            if (!tileentitylectern.hasBook()) return; // Paper
+             LecternBlock.resetBookState(world, pos, state, true);
+             world.playSound((Player) null, pos, SoundEvents.BOOK_PUT, SoundSource.BLOCKS, 1.0F, 1.0F);
+             world.gameEvent(player, GameEvent.BLOCK_CHANGE, pos);
+diff --git a/src/main/java/net/minecraft/world/level/block/entity/LecternBlockEntity.java b/src/main/java/net/minecraft/world/level/block/entity/LecternBlockEntity.java
+index 5d0eb0dbfa1f6c2a8e49f8e7b7b0089d7e27764d..124c68ffcc4e9c968dadb6274a2ac5275e67799c 100644
+--- a/src/main/java/net/minecraft/world/level/block/entity/LecternBlockEntity.java
++++ b/src/main/java/net/minecraft/world/level/block/entity/LecternBlockEntity.java
+@@ -246,6 +246,17 @@ public class LecternBlockEntity extends BlockEntity implements Clearable, MenuPr
+         if (this.level instanceof ServerLevel && book.is(Items.WRITTEN_BOOK)) {
+             WrittenBookItem.resolveBookComponents(book, this.createCommandSourceStack(player), player);
+         }
++        // Paper start
++        if (player != null && getLevel() != null) {
++            io.papermc.paper.event.player.PlayerInsertLecternBookEvent event = new io.papermc.paper.event.player.PlayerInsertLecternBookEvent(
++                (org.bukkit.entity.Player) player.getBukkitEntity(),
++                (org.bukkit.block.Lectern) getLevel().getWorld().getBlockAt(worldPosition.getX(), worldPosition.getY(), worldPosition.getZ()).getState(false),
++                org.bukkit.craftbukkit.inventory.CraftItemStack.asCraftMirror(book)
++            );
++            if (!event.callEvent()) return ItemStack.EMPTY;
++            return ((org.bukkit.craftbukkit.inventory.CraftItemStack) event.getBook()).handle;
```

Comment:
Since there is a setter for the book itemstack on the event, you cannot cast it to CraftItemStack since there is no guarantee that it will be an instance of it.